File: BE/BE/src/main/java/DoAn/BE/project/entity/Issue.java
package DoAn.BE.project.entity;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

import jakarta.persistence.*;
import lombok.*;

import DoAn.BE.user.entity.User;

@Entity
@Table(name = "issues")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Issue {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "issue_id")
    private Long issueId;

    @ManyToOne
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @ManyToOne
    @JoinColumn(name = "sprint_id")
    private Sprint sprint;

    @Column(name = "issue_key", nullable = false, unique = true, length = 20)
    private String issueKey;  // VD: PROJ-001, PROJ-002

    @Column(nullable = false, length = 255)
    private String title;

    @Column(columnDefinition = "NVARCHAR(MAX)")
    private String description;


    @ManyToOne
    @JoinColumn(name = "status_id", nullable = false)
    private IssueStatus issueStatus;

    @Enumerated(EnumType.STRING)
    @Column(length = 20, nullable = false)
    private Priority priority = Priority.MEDIUM;

    @ManyToOne
    @JoinColumn(name = "reporter_id", nullable = false)
    private User reporter;  // Người tạo issue

    @ManyToOne
    @JoinColumn(name = "assignee_id")
    private User assignee;  // Người được giao việc

    @Column(name = "estimated_hours", precision = 5, scale = 2)
    private BigDecimal estimatedHours;

    @Column(name = "actual_hours", precision = 5, scale = 2)
    private BigDecimal actualHours;

    @Column(name = "due_date")
    private LocalDate dueDate;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;


    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Helper methods
    public boolean isOverdue() {
        return this.dueDate != null && 
               LocalDate.now().isAfter(this.dueDate) &&
               !isDone();
    }

    public boolean isDone() {
        return this.issueStatus != null && 
               this.issueStatus.getName() != null &&
               "Done".equals(this.issueStatus.getName());
    }

    public boolean isAssigned() {
        return this.assignee != null;
    }

    public void assignTo(User user) {
        this.assignee = user;
        this.updatedAt = LocalDateTime.now();
    }

    public void changeStatus(IssueStatus newStatus) {
        this.issueStatus = newStatus;
        this.updatedAt = LocalDateTime.now();
    }

    // Enum
    public enum Priority {
        LOW,       // Thấp
        MEDIUM,    // Trung bình
        HIGH,      // Cao
        CRITICAL   // Khẩn cấp
    }
}

File: mobile/lib/data/models/issue.dart
class IssueStatus {
  final int statusId;
  final String name;

  IssueStatus({required this.statusId, required this.name});

  factory IssueStatus.fromJson(Map<String, dynamic> json) {
    return IssueStatus(
      statusId: json['statusId'] ?? 0,
      name: json['name'] ?? 'Unknown',
    );
  }
}

class Comment {
  final int commentId;
  final String content;
  final String authorName;
  final String? authorAvatarUrl;
  final String createdAt;

  Comment({
    required this.commentId,
    required this.content,
    required this.authorName,
    this.authorAvatarUrl,
    required this.createdAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    String authorName = 'Unknown';
    String? avatarUrl;
    if (json['user'] != null) {
      authorName = json['user']['username'] ?? 'Unknown';
      avatarUrl = json['user']['avatarUrl'];
    } else if (json['author'] != null) {
      authorName = json['author']['username'] ?? 'Unknown';
      avatarUrl = json['author']['avatarUrl'];
    } else if (json['authorName'] != null) {
      authorName = json['authorName'];
    }

    return Comment(
      commentId: json['commentId'] ?? json['id'] ?? 0,
      content: json['content'] ?? '',
      authorName: authorName,
      authorAvatarUrl: avatarUrl,
      createdAt: json['createdAt'] ?? '',
    );
  }
}

class Issue {
  final int issueId;
  final String title;
  final String issueKey;
  final String description;
  final IssueStatus status;
  final String priority;
  final String? projectName;
  final String? dueDate;
  final String? assigneeName;
  final String? assigneeAvatarUrl;
  final String? reporterName;
  final double? estimatedHours;
  final double? actualHours;
  final bool isOverdue;
  final List<Comment> comments;
  final int? projectId;
  final int? sprintId;

  Issue({
    required this.issueId,
    required this.title,
    required this.issueKey,
    required this.description,
    required this.status,
    required this.priority,
    this.projectName,
    this.dueDate,
    this.assigneeName,
    this.assigneeAvatarUrl,
    this.reporterName,
    this.estimatedHours,
    this.actualHours,
    this.isOverdue = false,
    this.comments = const [],
    this.projectId,
    this.sprintId,
  });

  factory Issue.fromJson(Map<String, dynamic> json) {
    // Parse status - can be object or direct fields (statusName, statusId, statusColor)
    IssueStatus status;
    if (json['status'] != null && json['status'] is Map) {
      status = IssueStatus.fromJson(json['status']);
    } else if (json['issueStatus'] != null && json['issueStatus'] is Map) {
      status = IssueStatus.fromJson(json['issueStatus']);
    } else if (json['statusName'] != null) {
      // API returns direct fields: statusName, statusId, statusColor
      status = IssueStatus(
        statusId: json['statusId'] ?? 0,
        name: json['statusName'] ?? 'Unknown',
      );
    } else {
      // Fallback
      String statusName = json['issueStatus'] ?? json['status'] ?? 'Unknown';
      status = IssueStatus(statusId: 0, name: statusName.toString());
    }

    // Parse comments
    List<Comment> commentsList = [];
    if (json['comments'] != null && json['comments'] is List) {
      commentsList = (json['comments'] as List).map((e) => Comment.fromJson(e)).toList();
    }

    // Parse assignee
    String? assigneeName;
    String? assigneeAvatarUrl;
    if (json['assignee'] != null) {
      assigneeName = json['assignee']['username'];
      assigneeAvatarUrl = json['assignee']['avatarUrl'];
    }

    // Parse reporter
    String? reporterName;
    if (json['reporterName'] != null) {
      reporterName = json['reporterName'];
    } else if (json['reporter'] != null) {
      reporterName = json['reporter']['username'];
    }

    return Issue(
      issueId: json['issueId'] ?? 0,
      title: json['title'] ?? '',
      issueKey: json['issueKey'] ?? '',
      description: json['description'] ?? '',
      status: status,
      priority: json['priority'] ?? 'MEDIUM',
      projectName: json['projectName'] ?? json['project']?['name'],
      dueDate: json['dueDate'],
      assigneeName: assigneeName,
      assigneeAvatarUrl: assigneeAvatarUrl,
      reporterName: reporterName,
      estimatedHours: json['estimatedHours']?.toDouble(),
      actualHours: json['actualHours']?.toDouble(),
      isOverdue: json['isOverdue'] ?? false,
      comments: commentsList,
      projectId: json['project']?['projectId'] ?? json['projectId'],
      sprintId: json['sprint']?['sprintId'] ?? json['sprintId'],
    );
  }

  // Helper to get status name for UI
  String get statusName => status.name;

  Issue copyWith({
    int? issueId,
    String? title,
    String? issueKey,
    String? description,
    IssueStatus? status,
    String? priority,
    String? projectName,
    String? dueDate,
    String? assigneeName,
    String? assigneeAvatarUrl,
    String? reporterName,
    double? estimatedHours,
    double? actualHours,
    bool? isOverdue,
    List<Comment>? comments,
    int? projectId,
    int? sprintId,
  }) {
    return Issue(
      issueId: issueId ?? this.issueId,
      title: title ?? this.title,
      issueKey: issueKey ?? this.issueKey,
      description: description ?? this.description,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      projectName: projectName ?? this.projectName,
      dueDate: dueDate ?? this.dueDate,
      assigneeName: assigneeName ?? this.assigneeName,
      assigneeAvatarUrl: assigneeAvatarUrl ?? this.assigneeAvatarUrl,
      reporterName: reporterName ?? this.reporterName,
      estimatedHours: estimatedHours ?? this.estimatedHours,
      actualHours: actualHours ?? this.actualHours,
      isOverdue: isOverdue ?? this.isOverdue,
      comments: comments ?? this.comments,
      projectId: projectId ?? this.projectId,
      sprintId: sprintId ?? this.sprintId,
    );
  }
}
